### DISCUSSION TITLE
Digest authentication support

### DISCUSSION THREAD

**Discussion by tahdsh** [grafana/grafana-infinity-datasource#309] [Q&A] [answered]
Hi guys,
What if the http request need authentication to access data, how can we visualise this data with grafana ?

**Comment by yesoreyeram** [comment]
You can setup the authentication in datasource config. Refer [docs](https://sriramajeyam.com/grafana-infinity-datasource/wiki/authentication/) for more details.

**Reply by tahdsh** [reply]
Hello, thanks a lot for the answer.
Im trying the online version of grafana infinity, how can i add a basic authentication to the http header ?
If you can give me steps to this or an example, i'm new to using json rest api.

**Reply by yesoreyeram** [reply]
What do you mean by online version? Heroku instance is just for demo.

**Comment by tahdsh** [comment]
Hello,
Im trying to pull metrics from hpe imc with rest api into grafana
Im using infinty datasource, but im getting a query error when sending http request.
I added a header to the request with basic authentication,
Header name is “Authorization:” and Header value is “Basic username:password” in base64 encoding
But i get a status 500 internal server error : query data error 
![image_2022-05-11_153251359](https://user-images.githubusercontent.com/105001058/167862791-c2e0bedd-d339-473b-907e-e624b3468314.png)

**Reply by tahdsh** [reply]
If you have an idea what might be wrong with my configuration

**Reply by yesoreyeram** [reply]
May be the URL is not reachable by grafana instance.. does grafana server and your endpoint runs in the same server?

**Reply by tahdsh** [reply]
Yes it is running on same server.

I tried the url with curl and i get this result even with adding user authentification.
![curl test](https://user-images.githubusercontent.com/105001058/168067318-9a7136a3-89ff-47b1-9480-436cd2e898b1.PNG)

I think its worth mentionning that in grafana when i dont add authentification header i get response code from server : 401. Error message : 401 Unaut
horized and when i add the header 
Header name is “Authorization:” and Header value is “Basic username:password” in base64 encoding
i get this 500 error

**Reply by tahdsh** [reply]
I found that AuthType 4  (Digest Auth ) is used in authentication to the hpe imc server ( the one that i want to pull metrics from)  now i can get the response with curl, but not sure if infinity datasource supports Digest Auth, or is there anyway to add digest parameters with the http request?

**Reply by yosiasz** [reply]
How about adding an Authorization header and then add that hash or also try it as user:password?

**Reply by tahdsh** [reply]
The problem is for each authentication the server generates a nonce, nonce count and cnonce that should be included in the response.

What we have here  : 

qop = "auth"

HA1 = MD5(username:realm:password)

HA2 = MD5(method:digestURI)

the response that should be sent to the server:

response = MD5(
HA1:nonce:nonceCount:cnonce:qop:HA2)

![digest auth](https://user-images.githubusercontent.com/105001058/168550866-62b8c543-733d-4a0e-bb56-dc0e63ddbb85.PNG)

**Comment by yesoreyeram** [comment, marked as answer]
Can you try v1.0.0-dev.3 version from https://github.com/yesoreyeram/grafana-infinity-datasource/releases/tag/v1.0.0-dev.3 which have basic digest auth support

**Reply by tahdsh** [reply]
Thanks a lot for integrating digest authentication mode, it works very well.
Truly great tool to present rest api data
